File: easy/0104_maximum_depth_of_binary_tree/solution.cpp
#include "../../utils/cpp/nodes.hpp"

class Solution {
public:
  int maxDepth(utils::TreeNode *root) {
    if (!root)
      return 0;

    return 1 + std::max(maxDepth(root->left), maxDepth(root->right));
  }
};

File: easy/0104_maximum_depth_of_binary_tree/solution.py
from utils.python.nodes import TreeNode


class Solution:
    def maxDepth(self, root: TreeNode | None) -> int:
        if not root:
            return 0

        stack = [(root, 1)]
        max_depth = -1

        while stack:
            node, depth = stack.pop()

            if depth > max_depth:
                max_depth = depth

            if node.left:
                stack.append((node.left, depth + 1))
            if node.right:
                stack.append((node.right, depth + 1))

        return max_depth
